Implements a way to run Migrations in a non-shell environment

HavokInspiration opened the pull request:
This PR aims to give the ability to run Migrations in a non-shell environment.
It implements a new ``Migrations`` class that exposes three main methods : ``status``, ``migrate`` and ``rollback``.

The constructor accepts an array that will allow users to define default options to run with the methods. The expected and allowed values are the one that are common between each command : 
```
     * - `connection` The datasource connection to use
     * - `source` The folder where migrations are in
     * - `plugin` The plugin containing the migrations
```

So this can be done :

```php
        $migrations = new Migrations([
            'connection' => 'customConnection',
            'source' => 'MyCustomDir'
        ]);
        
        // Returns an array with all migrations and their status
        $status = $migrations->status();

        // Returns a bool
        $migrate = $migrations->migrate();
        
        // If migrate is false, use the getLastError() method to get the exception message
        // that was raised
        $error = $migrations->getLastError();
```

As you can see, I don't have to pass the options array for each method. They can be overridden as well by using the $options parameter for each method :

```php
// Returns an array with all migrations and their status
$status = $migrations->status(['connection' => 'default']);
```

In order to implement this, I had to create a ``CakeManager`` class that overrides phinx's ``Manager`` class. This was needed to separate logic and output for the ``status`` command.
I also fixed a wrong behavior of the ``status`` command : if you run it with the ``format`` option set to ``json``, it will now only returns a JSON response.

**Possible issues :**
- if there is no migrations to run, the ``migrate()`` method will return ``true`` anyway. Same applies to ``rollback()``. Generally speaking, ``migrate()`` and ``rollback()`` will return ``false`` only if an exception is thrown while running the command.
- I did not implement the ``MarkMigrated`` and ``Create`` commands. I could see some value for ``MarkMigrated`` but none with ``Create``. I'm waiting for feedback on this one.
- tests might be failing, I was only able to partly run them with MySQL on my dev setup

**Possible improvements :**
- have the ``migrate()`` and ``rollback()`` method to return what migrations were migrated and rollbacked. This might need some tuning with the ``CakeManager`` to have something similar ``status()`` (split logic and output). I did not try it but I think this might be doable.

Docs will come up when everything will be settled and the implementation confirmed.

Refs #37

HavokInspiration commented on src/ConfigurationTrait.php:
```diff
@@ -46,11 +46,12 @@
+     * @return \Migrations\Phinx\Config\Config
      */
-    public function getConfig()
+    public function getConfig($forceRefresh = false)
```
Might need to be reverted. It was needed at some point but not anymore I think.

dereuromark commented on src/CakeManager.php:
```diff
@@ -0,0 +1,58 @@
+ * Overrides Phinx Manager class in order to provide an interface
+ * for running migrations within an app
+ */
+class CakeManager extends Manager {
```
Any reason you are using CakeManager and not Manager with aliasing of the original class?

HavokInspiration replied:
Well, we have ``CakeAdapter``, so I didn't really give this much thought.
I surely can remove the ``Cake`` prefix.

dereuromark replied:
No, not because of me - I was just wondering :)

HavokInspiration commented:
The SQLite failing tests are puzzling.
I don't really understand why they are happening.

Good news is it also happens on my dev setup... so working on this will be easier.

HavokInspiration commented on src/Migrations.php:
```diff
@@ -0,0 +1,206 @@
+    {
+        $this->setInput($input);
+        $config = $this->getConfig();
+        $manager = new CakeManager($config, $this->output);
```
Check if the same instance can be used and the ConfigInterface instance swapped with a method from the original phinx Manager class

HavokInspiration commented:
I just pushed new commits.

- I implemented the ``markMigrated`` command in the ``Migrations`` class. I had to change the implementation of the "real" command to get this done (something similar to the ``Status`` command where the logic and output are separated)
- I did some docblocks type cleanup and changed how the ``getConnectionName()`` method from the trait is returned. The current implementation did not use the parameter passed to the method.
- I updated the test to they pass on SQLite and MySQL

HavokInspiration commented:
Note to self : add tests where "default" params are overridden in a method call.

HavokInspiration commented:
Just pushed a new commit where I add more tests and where I added a new method ``resetMigrations()`` in the ``CakeManager`` to purge the migrations stored in the object when two following calls of the ``Migrations`` method are done with a change in the migrations files path.

lorenzo commented on src/CakeManager.php:
```diff
@@ -0,0 +1,136 @@
+     * Prints the specified environment's migration status.
+     *
+     * @param string $environment
+     * @param null $format
```
string ?

HavokInspiration replied:
``null | string`` actually. null being the default.

lorenzo commented on src/CakeManager.php:
```diff
@@ -0,0 +1,136 @@
+     *
+     * @param int|string $version Version number of the migration to check
+     * @param string $path Path where the migration file is located
+     * @return bool|string A string with the error message or true if success
```
when does this return a string?

HavokInspiration replied:
Never, it's a mistake. Probably a copy / paste from another method (from a previous version of code).

lorenzo commented on src/Command/Status.php:
```diff
@@ -33,4 +35,79 @@ protected function configure()
+     * @param OutputInterface $output
+     * @return void
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
```
Why did you have to re-implement this class?

HavokInspiration replied:
Because the original command has output statements in the ``Manager::printStatus()`` method.
So I moved the output in this method (because output should only be in Command classes in my opinion) to just have what was needed to get the status of migrations in the ``CakeManager::printStatus()`` method.

lorenzo replied:
I see

lorenzo commented on src/Migrations.php:
```diff
@@ -0,0 +1,280 @@
+        try {
+            $this->run('migrate', $params, $input);
+            return true;
+        } catch (\Exception $e) {
```
Would it be too bad to just let the exception bubble up? I think it is better than jus returning false

HavokInspiration replied:
Sure I can do that.

lorenzo commented on tests/TestCase/MigrationsTest.php:
```diff
@@ -0,0 +1,238 @@
+        if (in_array('phinxlog', $tables)) {
+            $ormTable = TableRegistry::get('phinxlog', ['connection' => $this->Connection]);
+            $this->Connection->execute(
+                $this->Connection->driver()->schemaDialect()->truncateTableSql($ormTable->schema())[0]
```
This is php 5.5 syntax

lorenzo commented on tests/TestCase/MigrationsTest.php:
```diff
@@ -0,0 +1,238 @@
+                $this->Connection->driver()->schemaDialect()->truncateTableSql($ormTable->schema())[0]
+            );
+        }
+
```
Extra new line

lorenzo commented:
Amazing jon @HavokInspiration :clap:

HavokInspiration commented:
I addressed the feedback (and rebased on top of master).
The ``Migrations`` class methods now let the exceptions bubble up. Regarding the tests, I had to expect ``Exception`` since with SQLite and MySQL, the exceptions raised are not the same.

lorenzo commented:
@HavokInspiration out of curiosity, what was the difference between the 2 exceptions?

HavokInspiration commented:
@lorenzo For a test, SQLite threw a ``InvalidArgumentException`` and MySQL a ``PDOException``.
My first attempt / tests were with SQLite locally and it threw :
```
2) Migrations\Test\MigrationsTest::testRollbackErrors
InvalidArgumentException: The specified column doesn't exist: radix
```

So I excepted a ``InvalidArgumentException`` in the tests.
When running the tests with MySQL :

```
1) Migrations\Test\MigrationsTest::testRollbackErrors
Failed asserting that exception of type "PDOException" matches expected exception "InvalidArgumentException"
```

lorenzo commented:
I see... I think this is good to go now :D

dereuromark commented:
:clap:
